lnbits/extensions/boltz/boltz.py

```diff
@@ -0,0 +1,319 @@
+    """
+
+    claim_privkey = ec.PrivateKey(os.urandom(32), True, net)
+    claim_pubkey_hex = hexlify(claim_privkey.sec()).decode("UTF-8")
+    preimage = os.urandom(32)
+    preimage_hash = sha256(preimage).hexdigest()
+
+    res = create_post_request(BOLTZ_URL + "/createswap", {
+        "type": "reversesubmarine",
+        "pairId": "BTC/BTC",
+        "orderSide": "buy",
+        "invoiceAmount": data.amount,
+        "preimageHash": preimage_hash,
+        "claimPublicKey": claim_pubkey_hex
+    })
+
+    swap = ReverseSubmarineSwap(
+        id = swap_id,
+        amount = data.amount,
+        wallet = data.wallet,
+        onchain_address = data.onchain_address,
+        instant_settlement = data.instant_settlement,
+        claim_privkey = claim_privkey.wif(net),
+        preimage = preimage.hex(),
+        boltz_id = res["id"],
+        invoice = res["invoice"],
+        timeout_block_height = res["timeoutBlockHeight"],
+        lockup_address = res["lockupAddress"],
+        onchain_amount = res["onchainAmount"],
+        redeem_script = res["redeemScript"],
+        time = getTimestamp(),
+    )
+
+    asyncio.ensure_future(wait_for_onchain_tx(swap))
+
+    # TODO: think about what happens if pay_invoice fails
+    # ensure_future is used because pay_invoice is stuck
+    # as long as boltz does not see the onchain tx
+    asyncio.ensure_future(pay_invoice(
+       wallet_id=data.wallet,
+       payment_request=res["invoice"],
+       description=f"reverse submarine swap for {data.amount} sats on boltz.exchange",
+       extra={"tag": "boltz"},
+    ))
+
+    return swap
+
+def listen_to_stream_swap_status(swap: ReverseSubmarineSwap):
+    messages = SSEClient(BOLTZ_URL + "/streamswapstatus?id=" + swap.boltz_id)
+    for msg in messages:
+        data = json.loads(msg.data)
+        if swap.instant_settlement == True:
+            status = "transaction.mempool"
+        else:
+            status = "transaction.confirmed"
+        if data["status"] == status:
+            return swap
+        if  data["status"] == "transaction.failed":
+            raise HTTPException(
+                status_code=HTTPStatus.INTERNAL_SERVER_ERROR, detail="boltz status transaction.failed! no onchain funds at boltz"
+            )
+        if  data["status"] == "invoice.expired":
+            raise HTTPException(
+                status_code=HTTPStatus.INTERNAL_SERVER_ERROR, detail="ln invoice has expired!"
+            )
+
+async def wait_for_onchain_tx(swap: Union[ReverseSubmarineSwap, SubmarineSwap]):
+    loop = asyncio.get_running_loop()
+    try:
+        swap = await loop.run_in_executor(None, listen_to_stream_swap_status, swap)
+    except HTTPException:
+        # TODO: what todo?!
+        print("listen_to_stream_swap_status failed")
+    tx = await create_onchain_tx(swap)
+    return await send_onchain_tx(tx)
+
+def get_mempool_tx(address):
+    txs = create_get_request(MEMPOOL_SPACE_URL + f"/api/address/{address}/txs")
```

Comment:
I like the idea of using mempool.space to fetch onchain data instead of relying just on the Boltz API. But there is one catch with that architecture. Albeit it not happening in our regtest setup, there is a race condition between the boltz and mempool.space API. Like for example boltz sending you the `transaction.mempool` event before that transaction arrives in the mempool of the mempool.space instance. This is actually very likely to happen on mainnet.

Reply:
Whats the best fix for this?

Reply:
It's tricky. One option is listening to the websocket of the mempool.space instance for transactions that involve the lockup address of the swap. And having those events trigger claim transactions.

Reply:
we could keep retrying mempool.space a couple of times, after boltz sends the mempool event, it should propagate in a reasonable amount of time. imho, it should be fine.

Reply:
So we have two options:
- either periodically retrying mempool.space REST API after Boltz sends the event
- or [using the mempool.space WebSocket](https://mempool.space/docs/api/websocket)

Both should be fine

Reply:
cool, didnt know about the mempool.space websocket, i guess thats the better option?